File: src/updatable_encryption_python/encrypt_message.py
import numpy as np

#from parameters import *
from updatable_encryption_python.utility_functions import *
from updatable_encryption_python.parameters import *


def compact_error(e_vec):
    qhalf = round(q/2)
    for i in range(len(e_vec)):
        if e_vec[i] > qhalf:
            e_vec[i] = e_vec[i] - q


# generate error vectors
def generate_error_vector(m_tilda, nk, alpha, q, sigma, message_blocks):
    e0 = sample_normal_matrix(1, m_tilda, 0, edisp)[0]
    VerifyE(e0)
    #d = calculate_d(e0, m_tilda, alpha, q, sigma)
    e1 = sample_normal_matrix(1, nk, 0, edisp)[0]
    VerifyE(e1)
    e2 = sample_normal_matrix(1, message_blocks * nk, 0, edisp)[0]
    VerifyE(e2)
    #parameters.updateError(e0, e1, e2)
    return (e0, e1, e2)
    #return (np.zeros(m_tilda), np.zeros(nk), np.zeros(nk))

# calculate d    
def calculate_d(e0, m_tilda, alpha, q, sigma):
    sum = 0
    for value in e0:
        sum+= value**2
    #print("sum=", sum, "m_tilda=", m_tilda*((alpha*q)**2), "sigma=", sigma**2)
    d = np.sqrt(sum + m_tilda*((alpha*q)**2) * (sigma**2))    
    return d

# encrypted an encoded message
def Encrypt(pk, encoded_message, H_input, updateMatrix, updateWithMatrix, message_size):
    global s_hidden, e0_hidden, e1_hidden, e2_hidden
    A0, A1, A2 = pk

    # H_input is None for new encryption
    # H_input is not None when used in token_generation for encrypting the 0-message
    if not H_input is None:
        new_Hmu = H_input
    else:
        new_Hmu = generate_invertible_matrixQ(n)

    # Combined matrix
    A_mu = np.block([A0, (A1 + np.dot(new_Hmu, G)) % q])  

    # calculate the number of needed blocks
    message_blocks = message_size // nk
    if message_size % nk != 0:
        message_blocks+=1

    # add A2 matrixes depending on message length
    for _ in range(message_blocks):
        A_mu =  np.block([A_mu, A2])   

    # Random vector
    s = np.random.randint(0, q, n) 
    updateS(s)

    #TO DO choose value of alpha
    e0, e1, e2 = generate_error_vector(m_tilda, nk, alpha, q, edisp, message_blocks)
    


    #print("----------------------------------- e0=", e0)
    #print("----------------------------------- e1=", e1)
    #print("----------------------------------- e2=", e2)
    if updateWithMatrix:
        updateError2(e0, e1, e2, updateMatrix)
    else:
        updateError(e0, e1, e2)
    
    #print("e0=", e0)
    #print("e1=", e1)
    #print("e2=", e2)
    #print("encoded msg: ", encode(message))
    b = (s @ A_mu + np.concatenate((e0, e1, e2)) + np.concatenate((np.zeros(m_tilda), np.zeros(nk), encoded_message))).astype(int) % q 

    return (new_Hmu, b)

File: chaincode-go/updatable_encryption/encrypt_message.go
package updatable_encryption

import (
	"fmt"
)

/*
func compact_error(e_vec []int) []int {
	qhalf := math.Round(float64(C.q) / 2)

	for i := range e_vec {
		if e_vec[i] > int(qhalf) {
			e_vec[i] = e_vec[i] - C.q
		}
	}

	return e_vec
}


func compact_error2(e_vec []int) []int {
	qhalf := math.Round(float64(C.q) / 2)

	for i := range e_vec {
		if e_vec[i] > int(qhalf) {
			e_vec[i] = e_vec[i] - C.q
		}
	}

	for i := range e_vec {
		if e_vec[i] < -int(qhalf) {
			e_vec[i] = e_vec[i] + C.q
		}
	}

	return e_vec
}
*/

func generate_error_vector() ([]int, []int, []int) {
	e0 := sample_normal_matrix(1, C.m_tilda, 0, C.edisp)[0]
	VerifyE(e0)
	//#d = calculate_d(e0, m_tilda, alpha, q, sigma)
	e1 := sample_normal_matrix(1, C.nk, 0, C.edisp)[0]
	VerifyE(e1)
	e2 := sample_normal_matrix(1, C.nk, 0, C.edisp)[0]
	VerifyE(e2)

	return e0, e1, e2
}

func generate_error_vector_multiple(size int) ([]int, []int, []int) {
	e0 := sample_normal_matrix(1, C.m_tilda, 0, C.edisp)[0]
	VerifyE(e0)
	//#d = calculate_d(e0, m_tilda, alpha, q, sigma)
	e1 := sample_normal_matrix(1, C.nk, 0, C.edisp)[0]
	VerifyE(e1)
	e2 := sample_normal_matrix(1, size, 0, C.edisp)[0]
	VerifyE(e2)

	return e0, e1, e2
}

/*
# calculate d
def calculate_d(e0, m_tilda, alpha, q, sigma):
    sum = 0
    for value in e0:
        sum+= value**2
    #print("sum=", sum, "m_tilda=", m_tilda*((alpha*q)**2), "sigma=", sigma**2)
    d = np.sqrt(sum + m_tilda*((alpha*q)**2) * (sigma**2))
    return d
*/

func Encrypt(A0, A1, A2 [][]int, encoded_message []int, H_mu_input [][]int, updateMatrix [][]int, updateWithMatrix bool) ([][]int, []int, error) {

	var H_mu [][]int

	if H_mu_input != nil {
		H_mu = H_mu_input
	} else {
		H_mu = generate_invertible_matrixQ(C.n)
	}

	if H_mu == nil {
		return nil, nil, fmt.Errorf("encrypt - H_mu")
	}

	HmG, err1 := dot_product_MM(H_mu, C.G)

	if err1 != nil {
		return nil, nil, fmt.Errorf("encrypt - err1")
	}

	helper1, err2 := add_matrix(A1, HmG)

	if err2 != nil {
		return nil, nil, fmt.Errorf("encrypt - err2")
	}

	A_mu, err3 := concatenate_matrices_row(A0, helper1)

	if err3 != nil {
		return nil, nil, fmt.Errorf("encrypt - err3")
	}

	A_mu, err4 := concatenate_matrices_row(A_mu, A2)

	if err4 != nil {
		return nil, nil, fmt.Errorf("encrypt - err4")
	}

	s := sample_random_vector(C.n, C.q)

	err_s := updateS(s)

	if err_s != nil {
		return nil, nil, fmt.Errorf("encrypt - err_s")
	}

	e0, e1, e2 := generate_error_vector()

	if updateWithMatrix {
		err_e1 := update_error2(e0, e1, e2, updateMatrix)
		fmt.Printf("We are here!")
		if err_e1 != nil {
			return nil, nil, fmt.Errorf("encrypt - errUpdate1")
		}
	} else {
		err_e2 := update_error(e0, e1, e2)

		if err_e2 != nil {
			return nil, nil, fmt.Errorf("encrypt - errUpdate2")
		}
	}

	helper2, err5 := dot_product_VM(s, A_mu)

	if err5 != nil {
		return nil, nil, fmt.Errorf("encrypt - err5")
	}

	helper3 := concatenate_vector(concatenate_vector(e0, e1), e2)
	helper4 := concatenate_vector(concatenate_vector(make([]int, C.m_tilda), make([]int, C.nk)), encoded_message)

	b, err6 := add_vectors(helper2, helper3)

	if err6 != nil {
		return nil, nil, fmt.Errorf("encrypt - err6")
	}

	b, err7 := add_vectors(b, helper4)

	if err7 != nil {
		return nil, nil, fmt.Errorf("encrypt - err6")
	}

	return H_mu, b, nil
}

func Encrypt_multiple(A0, A1, A2 [][]int, encoded_message []int, H_mu_input [][]int, size int) ([][]int, []int, error) {

	var H_mu [][]int

	if H_mu_input != nil {
		H_mu = H_mu_input
	} else {
		H_mu = generate_invertible_matrixQ(C.n)
	}

	if H_mu == nil {
		return nil, nil, fmt.Errorf("encrypt - H_mu")
	}

	HmG, err1 := dot_product_MM(H_mu, C.G)

	if err1 != nil {
		return nil, nil, fmt.Errorf("encrypt - err1")
	}

	helper1, err2 := add_matrix(A1, HmG)

	if err2 != nil {
		return nil, nil, fmt.Errorf("encrypt - err2")
	}

	A_mu, err3 := concatenate_matrices_row(A0, helper1)

	if err3 != nil {
		return nil, nil, fmt.Errorf("encrypt - err3")
	}

	full_blocks := size / C.nk
	if size%C.nk != 0 {
		full_blocks += 1
	}

	for range full_blocks {
		A_mu, _ = concatenate_matrices_row(A_mu, A2)

		/*
			if err4 != nil {
				return nil, nil, fmt.Errorf("encrypt - err4")
			}
		*/
	}

	s := sample_random_vector(C.n, C.q)

	err_s := updateS(s)

	if err_s != nil {
		return nil, nil, fmt.Errorf("encrypt - err_s")
	}

	e0, e1, e2 := generate_error_vector_multiple(size)

	helper2, err5 := dot_product_VM(s, A_mu)

	if err5 != nil {
		return nil, nil, fmt.Errorf("encrypt - err5")
	}

	helper3 := concatenate_vector(concatenate_vector(e0, e1), e2)
	helper4 := concatenate_vector(concatenate_vector(make([]int, C.m_tilda), make([]int, C.nk)), encoded_message)

	b, err6 := add_vectors(helper2, helper3)

	if err6 != nil {
		return nil, nil, fmt.Errorf("encrypt - err6")
	}

	b, err7 := add_vectors(b, helper4)

	if err7 != nil {
		return nil, nil, fmt.Errorf("encrypt - err6")
	}

	return H_mu, b, nil
}
